secondary/docs/configuration.rst
2i configuration parameters
===========================

**maxVbuckets** (int)
    number of vbuckets configured in KV

**endpoint.dataport.bufferSize** (int)
    number of entries to buffer before flushing it, where each entry is for a vbucket's set of mutations that was flushed by the endpoint.

**endpoint.dataport.bufferTimeout** (int)
    timeout in milliseconds, to flush vbucket-mutations from endpoint

**endpoint.dataport.harakiriTimeout** (int)
    timeout in milliseconds, after which endpoint will commit harakiri if not activity

**endpoint.dataport.keyChanSize** (int)
    channel size of dataport endpoints data input

**endpoint.dataport.maxPayload** (int)
    maximum payload length, in bytes, for transmission data from router to downstream client

**endpoint.dataport.remoteBlock** (bool)
    should dataport endpoint block when remote is slow ?

**indexer.adminPort** (string)
    port for index ddl and status operations

**indexer.clusterAddr** (string)
    Local cluster manager address

**indexer.enableManager** (bool)
    Enable index manager

**indexer.numVbuckets** (int)
    Number of vbuckets

**indexer.scanPort** (string)
    port for index scan operations

**indexer.scanTimeout** (int)
    timeout, in milliseconds, timeout for index scan processing

**indexer.storage_dir** (string)
    Index file storage directory

**indexer.streamCatchupPort** (string)
    port for catchup stream

**indexer.streamInitPort** (string)
    port for inital build stream

**indexer.streamMaintPort** (string)
    port for maintenance stream

**indexer.compaction.interval** (int)
    Compaction poll interval in seconds

**indexer.compaction.minFrag** (int)
    Compaction fragmentation threshold percentage

**indexer.compaction.minSize** (uint64)
    Compaction min file size

**log.level** (string)
    logging level for the system. allowable values are silent|fatal|error|warn|info|debug|trace.

**projector.clusterAddr** (string)
    KV cluster's address to be used by projector

**projector.colocate** (bool)
    Whether projector will be colocated with KV. In which case `kvaddrs` specified above will be discarded

**projector.feedChanSize** (int)
    channel size for feed's control path and back path.

**projector.feedWaitStreamEndTimeout** (int)
    timeout, in milliseconds, to await a response for StreamEnd

**projector.feedWaitStreamReqTimeout** (int)
    timeout, in milliseconds, to await a response for StreamRequest

**projector.kvAddrs** (string)
    Comma separated list of KV-address to read mutations, this need to exactly match with KV-node's configured address

**projector.mutationChanSize** (int)
    channel size of projector's data path routine

**projector.name** (string)
    human readable name for this projector

**projector.routerEndpointFactory** (common.RouterEndpointFactory)
    RouterEndpointFactory callback to generate endpoint instances to push data to downstream

**projector.vbucketSyncTimeout** (int)
    timeout, in milliseconds, for sending periodic Sync messages.

**projector.adminport.listenAddr** (string)
    projector's adminport address listen for request.

**projector.adminport.maxHeaderBytes** (int)
    in bytes, is max. length of adminport http header used by projector

**projector.adminport.name** (string)
    human readable name for this adminport, must be supplied

**projector.adminport.readTimeout** (int)
    timeout in milliseconds, is read timeout for adminport http server used by projector

**projector.adminport.urlPrefix** (string)
    url prefix (script-path) for adminport used by projector

**projector.adminport.writeTimeout** (int)
    timeout in milliseconds, is write timeout for adminport http server used by projector

**projector.client.exponentialBackoff** (int)
    multiplying factor on retryInterval for every attempt with server

**projector.client.maxRetries** (int)
    maximum number of timest to retry

**projector.client.retryInterval** (int)
    retryInterval, in milliseconds, when connection refused by server

**projector.client.urlPrefix** (string)
    url prefix (script-path) for adminport used by projector

**projector.dataport.indexer.genServerChanSize** (int)
    request channel size of indexer dataport's gen-server routine

**projector.dataport.indexer.maxPayload** (int)
    maximum payload length, in bytes, for receiving data from router

**projector.dataport.indexer.tcpReadDeadline** (int)
    timeout, in milliseconds, while reading from socket

**queryport.client.connPoolAvailWaitTimeout** (int)
    timeout, in milliseconds, to wait for an existing connection from the pool before considering the creation of a new one

**queryport.client.connPoolTimeout** (int)
    timeout, in milliseconds, is timeout for retrieving a connection from the pool

**queryport.client.maxPayload** (int)
    maximum payload, in bytes, for receiving data from server

**queryport.client.poolOverflow** (int)
    maximum number of connections in a pool

**queryport.client.poolSize** (int)
    number simultaneous active connections connections in a pool

**queryport.client.readDeadline** (int)
    timeout, in milliseconds, is timeout while reading from socket

**queryport.client.writeDeadline** (int)
    timeout, in milliseconds, is timeout while writing to socket

**queryport.indexer.maxPayload** (int)
    maximum payload, in bytes, for receiving data from client

**queryport.indexer.pageSize** (int)
    number of index-entries that shall be returned as single payload

**queryport.indexer.readDeadline** (int)
    timeout, in milliseconds, is timeout while reading from socket

**queryport.indexer.streamChanSize** (int)
    size of the buffered channels used to stream request and response.

**queryport.indexer.writeDeadline** (int)
    timeout, in milliseconds, is timeout while writing to socket


secondary/docs/settings_stats.md
## Indes Status

### Status can be obtained from:

    $ curl localhost:9102/getIndexStatus
    {
      "code": "success",
      "status": [
        {
            "defnId": 15184130463866081000,
            "name": "breweries",
            "bucket": "beer-sample",
            "secExprs": [ "`name`" ],
            "status": "Ready",
            "hosts": [ "127.0.0.1:8091" ]
        },
        {
        
            "defnId": 16833079877526070000,
            "name": "#primary",
            "bucket": "beer-sample",
            "isPrimary": true,
            "status": "Ready",
            "hosts": [ "127.0.0.1:8091" ]
        },
      ]
    }

* defnId     = Internal id of the index
* name       = User supplied name
* bucket     = The bucket the index is defined on
* secExpres  = Expressions defining the index fields (ordered)
* status     = Index status (Not Available / Created / Build / Ready)
* isPrimary  = Indicates if index is primary (index on document key)
* host       = Indexer node serving this index

              
## Index Statistics

### Stats can be obtained from:

        $ curl localhost:9102/stats
        {
           "default:num_mutations_queued" : 0,
           "default:first_name:scan_bytes_read" : 0,
           "default:first_name:avg_scan_latency" : 0,
           "default:first_name:total_scan_duration" : 0,
           "default:first_name:get_bytes" : 0,
           "default:first_name:num_docs_queued" : 0,
           "default:first_name:num_compactions" : 0,
           "memory_quota" : 268435456,
           "needs_restart" : false,
           "memory_used" : 823296,
           "default:first_name:num_docs_pending" : 0,
           "default:first_name:scan_wait_duration" : 0,
           "default:first_name:delete_bytes" : 0,
           "default:first_name:flush_queue_size" : 0,
           "default:first_name:num_flush_queued" : 0,
           "default:first_name:num_rows_returned" : 0,
           "default:first_name:avg_ts_interval" : 0,
           "default:mutation_queue_size" : 0,
           "default:first_name:num_snapshots" : 0,
           "default:first_name:disk_size" : 20480,
           "default:first_name:insert_bytes" : 0,
           "default:first_name:items_count" : 0,
           "default:first_name:data_size" : 0,
           "num_connections" : 0,
           "default:first_name:num_commits" : 0,
           "default:first_name:avg_scan_wait_latency" : 0,
           "default:first_name:build_progress" : 100,
           "default:first_name:num_requests" : 0,
           "default:first_name:num_docs_indexed" : 0
        }

### Per Index Statistics

For each index following set of stats will be published with bucket:index\_name:stats\_name format:

##### "default:first\_name32:num\_requests" : 4,
Number of scan requests served by the indexer

##### "default:first\_name:flush\_queue\_size" : 0,
Number of items currently in the flusher queue

##### "default:first\_name:num\_flush\_queued" : 0,
Number of items inserted into flusher queue so far

##### "default:first\_name:avg\_scan\_latency" : 0,
Mean scan latency

##### "default:first\_name:avg\_scan\_wait\_latency" : 0,
Mean latency incurred in waiting for read snapshot

##### "default:first\_name:num\_compactions" : 0,
Number of compactions performed the indexer

##### "default:first\_name:num\_compactions" : 0,
Number of database commits performed by the indexer

##### "default:first\_name:num\_snapshots" : 0,
Number of storage snapshots generated

##### "default:first\_name:avg\_ts\_interval" : 0,
Average interval for snapshot generation

##### "default:first\_name32:num\_docs\_pending" : 0,
Number of documents pending to be indexed


##### "default:first\_name32:num\_rows\_returned" : 60,
Total number rows returned so far by the indexer


##### "default:first\_name32:num\_docs\_queued" : 0,
Number of documents queued in indexer, but not indexed yet


##### "default:first\_name32:num\_docs\_indexed" : 62308,
Total number of docs indexed so far by the indexer


##### "default:first\_name32:disk\_size" : 13127680,
Total disk file size consumed by the index


##### "default:first\_name32:data\_size" : 13009634
Actual data size consumed by the index


##### "default:first\_name32:scan\_bytes\_read" : 4096
Total decoded bytes read from the index for serving scan requests


##### "default:first\_name32:items\_count" : 12000
Current total number of rows in the index. This is an approximate count and may be incorrect.

##### "default:first\_name7:build\_progress: 93
Initial build progress for the index. When index has completed initial build,
this stat will disappear from the stats.


##### "default:first\_name7:get\_bytes" : 0,
##### "default:first\_name7:delete\_bytes" : 0,
##### "default:first\_name7:insert\_bytes" : 1271211,
Bytes requested to forestdb operations (insert, delete, get)

### Indexer wide statistics

##### "num\_connections" : 0,
Current num connections used by indexer


##### "needs\_restart" : false,
Whether indexer needs restart to apply current mem\_quota settings


##### "default:first\_name7:scan\_wait\_duration" : 331286,
Total time taken (ns) for readable snapshot to be available for consistent query (similar to stale=false)


##### "default:first\_name7:total\_scan\_duration" : 1693478681,
Time taken for index scans

##### "default:num\_mutations\_queued" : 1000,
Total mutations queued so far

#### "default:mutation\_queue\_size": 10,
Current number of items in the mutations queue

#### "memory\_quota" : 256000,
Memory quota set for the indexer

#### "memory\_used" : 1000,
Memory used by the storage engine


## Indexer settings

All normal setings get/set should go through [metakv](https://github.com/couchbase/cbauth/tree/master/metakv).

Indexer will be able to dynamically update all the settings during runtime except memory\_quota.

#### Snapshot intervals (ms)

##### "settings.inmemory\_snapshot.interval" : 204800
##### "settings.persisted\_snapshot.interval" : 204800

Persisted snapshot intervals should be multiple of inmemory snapshot interval

Minimum: 100ms

#### Compaction settings


##### "settings.compaction.min\_size" : 1048576,

Minimum file size required to trigger compaction

Minimum: 50MB


##### "settings.compaction.check\_period" : 60,

Intervals to check if compaction threshold has reached

Minimum: 60s


##### "settings.compaction.min\_frag" : 30,

Minimum fragmentation percentage to trigger compaction

Minimum: 5%

#### System usage settings


##### "settings.compaction.interval" : "00:00,00:00",

Time interval during which compaction should be allowed to run (start\_hr:start\_min, end\_hr:end\_min)

Default value "00:00,00:00" means anytime.


##### "settings.max\_cpu\_percent" : 300,

Minimum: 100%

If it is set to 0, indexer will use all cpus


##### "settings.memory\_quota" : 0,

Indexer memory quota for forestdb.

Minimum: 0 (Disabled)


##### "settings.recovery.max\_rollbacks" : 5

Maximum number of snapshots to be kept even after compaction on forestdb

Minimum: 1

### Debugging Settings
All normal get/set should go through [metakv](https://github.com/couchbase/cbauth/tree/master/metakv).

For reading settings for debugging:

        $ curl localhost:9102/settings
        {
           "settings:compaction:min_frag" : 30,
           "settings:persisted_snapshot:interval" : 200,
           "settings:inmemory_snapshot:interval" : 200,
           "settings:max_cpu_percent" : 100,
           "settings:compaction:check_period" : 60,
           "settings:compaction:min_size" : 1048576,
           "settings:memory_quota" : 0,
           "settings:recovery:max_rollbacks" : 5,
           "settings:compaction:interval" : "00:00,00:00"
        }
        
To writing settings for debugging, read settings into a file, modify and

        curl localhost:9102/settings -d @settings:json


### Manual compaction
        $ curl localhost:9102/triggerCompaction



secondary/common/config.go
// Config is key, value map for system level and component configuration.
// Key is a string and represents a config parameter, and corresponding
// value is an interface{} that can be consumed using accessor methods
// based on the context of config-value.
//
// Config maps are immutable and newer versions can be created using accessor
// methods.
//
// Shape of config-parameter, the key string, is sequence of alpha-numeric
// characters separated by one or more '.' , eg,
//      "projector.adminport.readtimeout"

package common

import "encoding/json"
import "strings"
import "fmt"
import "reflect"
import "errors"
import "github.com/couchbase/indexing/secondary/logging"
import "github.com/couchbase/indexing/secondary/platform"
import "unsafe"
import "runtime"

// formula to compute the default CPU allocation for projector.
var projector_maxCpuPercent = (1 + (runtime.NumCPU() / 6)) * 100

// Threadsafe config holder object
type ConfigHolder struct {
	ptr unsafe.Pointer
}

func (h *ConfigHolder) Store(conf Config) {
	platform.StorePointer(&h.ptr, unsafe.Pointer(&conf))
}

func (h *ConfigHolder) Load() Config {
	confptr := platform.LoadPointer(&h.ptr)
	return *(*Config)(confptr)
}

// Config is a key, value map with key always being a string
// represents a config-parameter.
type Config map[string]ConfigValue

// ConfigValue for each parameter.
type ConfigValue struct {
	Value      interface{}
	Help       string
	DefaultVal interface{}
	Immutable  bool
}

// SystemConfig is default configuration for system and components.
// configuration parameters follow flat namespacing like,
//      "maxVbuckets"  for system-level config parameter
//      "projector.xxx" for projector component.
//      "projector.adminport.xxx" for adminport under projector component.
// etc...
var SystemConfig = Config{
	// system parameters
	"maxVbuckets": ConfigValue{
		1024,
		"number of vbuckets configured in KV",
		1024,
		true, // immutable
	},
	// projector parameters
	"projector.name": ConfigValue{
		"projector",
		"human readable name for this projector",
		"projector",
		true, // immutable
	},
	"projector.clusterAddr": ConfigValue{
		"localhost:9000",
		"KV cluster's address to be used by projector",
		"localhost:9000",
		true, // immutable
	},
	"projector.maxCpuPercent": ConfigValue{
		projector_maxCpuPercent,
		"Maximum percent of CPU that projector can use. " +
			"EG, 200% in 4-core (400%) machine would set indexer to " +
			"use 2 cores",
		projector_maxCpuPercent,
		false, // mutable
	},
	// Projector feed settings
	"projector.routerEndpointFactory": ConfigValue{
		RouterEndpointFactory(nil),
		"RouterEndpointFactory callback to generate endpoint instances " +
			"to push data to downstream",
		RouterEndpointFactory(nil),
		true, // immutable
	},
	"projector.feedWaitStreamReqTimeout": ConfigValue{
		10 * 1000,
		"timeout, in milliseconds, to await a response for StreamRequest",
		10 * 1000,
		false, // mutable
	},
	"projector.feedWaitStreamEndTimeout": ConfigValue{
		10 * 1000,
		"timeout, in milliseconds, to await a response for StreamEnd",
		10 * 1000,
		false, // mutable
	},
	"projector.mutationChanSize": ConfigValue{
		100,
		"channel size of projector's vbucket workers, " +
			"changing this value does not affect existing feeds.",
		100,
		false, // mutable
	},
	"projector.feedChanSize": ConfigValue{
		100,
		"channel size for feed's control path, " +
			"changing this value does not affect existing feeds.",
		100,
		false, // mutable
	},
	"projector.backChanSize": ConfigValue{
		10000,
		"channel size of projector feed's back-channel, " +
			"changing this value does not affect existing feeds.",
		10000,
		false, // mutable
	},
	"projector.syncTimeout": ConfigValue{
		2000,
		"timeout, in milliseconds, for sending periodic Sync messages, " +
			"changing this value does not affect existing feeds.",
		2000,
		false, // mutable
	},
	"projector.watchInterval": ConfigValue{
		5 * 60 * 1000, // 5 minutes
		"periodic tick, in milli-seconds to check for stale feeds, " +
			"a feed is considered stale when all its endpoint go stale.",
		5 * 60 * 1000,
		true, // immutable
	},
	"projector.cpuProfFname": ConfigValue{
		"",
		"filename to dump cpu-profile for projector.",
		"",
		false, // mutable
	},
	"projector.cpuProfile": ConfigValue{
		false,
		"boolean indicate whether to start or stop projector cpu profiling.",
		false,
		false, // mutable
	},
	"projector.memProfFname": ConfigValue{
		"",
		"filename to dump mem-profile for projector.",
		"",
		false, // mutable
	},
	"projector.memProfile": ConfigValue{
		false,
		"boolean to take current mem profile from projector.",
		false,
		false, // mutable
	},
	// projector dcp parameters
	"projector.dcp.genChanSize": ConfigValue{
		2048,
		"channel size for DCP's gen-server routine, " +
			"changing this value does not affect existing feeds.",
		2048,
		false, // mutable
	},
	"projector.dcp.dataChanSize": ConfigValue{
		10000,
		"channel size for DCP's data path routines, " +
			"changing this value does not affect existing feeds.",
		10000,
		false, // mutable
	},
	// projector adminport parameters
	"projector.adminport.name": ConfigValue{
		"projector.adminport",
		"human readable name for this adminport, must be supplied",
		"projector.adminport",
		true, // immutable

	},
	"projector.adminport.listenAddr": ConfigValue{
		"",
		"projector's adminport address listen for request.",
		"",
		true, // immutable
	},
	"projector.adminport.urlPrefix": ConfigValue{
		"/adminport/",
		"url prefix (script-path) for adminport used by projector",
		"/adminport/",
		true, // immutable
	},
	"projector.adminport.readTimeout": ConfigValue{
		0,
		"timeout in milliseconds, is http server's read timeout",
		0,
		true, // immutable
	},
	"projector.adminport.writeTimeout": ConfigValue{
		0,
		"timeout in milliseconds, is http server's write timeout",
		0,
		true, // immutable
	},
	"projector.adminport.maxHeaderBytes": ConfigValue{
		1 << 20, // 1 MegaByte
		"in bytes, is max. length of adminport http header",
		1 << 20, // 1 MegaByte
		true,    // immutable
	},
	// projector dataport client parameters
	"projector.dataport.remoteBlock": ConfigValue{
		true,
		"should dataport endpoint block when remote is slow, " +
			"does not affect existing feeds.",
		true,
		false, // mutable
	},
	"projector.dataport.keyChanSize": ConfigValue{
		100000,
		"channel size of dataport endpoints data input, " +
			"does not affect existing feeds.",
		100000,
		true, // immutable
	},
	"projector.dataport.bufferSize": ConfigValue{
		100,
		"number of entries to buffer before flushing it, where each entry " +
			"is for a vbucket's set of mutations that was flushed, " +
			"by the endpoint, does not affect existing feeds.",
		100,
		false, // mutable
	},
	"projector.dataport.bufferTimeout": ConfigValue{
		25,
		"timeout in milliseconds, to flush vbucket-mutations from, " +
			"endpoint, does not affect existing feeds.",
		25,    // 25ms
		false, // mutable
	},
	"projector.dataport.harakiriTimeout": ConfigValue{
		10 * 1000,
		"timeout in milliseconds, after which endpoint will commit harakiri " +
			"if not activity, does not affect existing feeds.",
		10 * 1000, //10s
		false,     // mutable
	},
	"projector.dataport.maxPayload": ConfigValue{
		1024 * 1024,
		"maximum payload length, in bytes, for transmission data from " +
			"router to downstream client, does not affect eixting feeds.",
		1024 * 1024, // 1MB
		true,        // immutable
	},
	// projector's adminport client, can be used by manager
	"manager.projectorclient.retryInterval": ConfigValue{
		16,
		"retryInterval, in milliseconds when connection refused by server",
		16,
		true, // immutable
	},
	"manager.projectorclient.maxRetries": ConfigValue{
		5,
		"maximum number of times to retry",
		5,
		true, // immutable
	},
	"manager.projectorclient.exponentialBackoff": ConfigValue{
		2,
		"multiplying factor on retryInterval for every attempt with server",
		2,
		true, // immutable
	},
	"manager.projectorclient.urlPrefix": ConfigValue{
		"/adminport/",
		"url prefix (script-path) for adminport used by projector",
		"/adminport/",
		true, // immutable
	},
	// indexer dataport parameters
	"indexer.dataport.genServerChanSize": ConfigValue{
		10000,
		"request channel size of indexer dataport's gen-server routine",
		10000,
		true, // immutable
	},
	"indexer.dataport.maxPayload": ConfigValue{
		1000 * 1024,
		"maximum payload length, in bytes, for receiving data from router",
		1000 * 1024, // bytes
		true,        // immutable
	},
	"indexer.dataport.tcpReadDeadline": ConfigValue{
		10 * 1000,
		"timeout, in milliseconds, while reading from socket",
		10 * 1000, // 10s
		true,      // immutable
	},
	// indexer queryport configuration
	"indexer.queryport.maxPayload": ConfigValue{
		1000 * 1024,
		"maximum payload, in bytes, for receiving data from client",
		1000 * 1024,
		true, // immutable
	},
	"indexer.queryport.readDeadline": ConfigValue{
		4000,
		"timeout, in milliseconds, is timeout while reading from socket",
		4000,
		true, // immutable
	},
	"indexer.queryport.writeDeadline": ConfigValue{
		4000,
		"timeout, in milliseconds, is timeout while writing to socket",
		4000,
		true, // immutable
	},
	"indexer.queryport.pageSize": ConfigValue{
		1,
		"number of index-entries that shall be returned as single payload",
		1,
		true, // immutable
	},
	"indexer.queryport.streamChanSize": ConfigValue{
		16,
		"size of the buffered channels used to stream request and response.",
		16,
		true, // immutable
	},
	// queryport client configuration
	"queryport.client.maxPayload": ConfigValue{
		1000 * 1024,
		"maximum payload, in bytes, for receiving data from server",
		1000 * 1024,
		true, // immutable
	},
	"queryport.client.readDeadline": ConfigValue{
		300000,
		"timeout, in milliseconds, is timeout while reading from socket",
		300000,
		true, // immutable
	},
	"queryport.client.writeDeadline": ConfigValue{
		4000,
		"timeout, in milliseconds, is timeout while writing to socket",
		4000,
		true, // immutable
	},
	"queryport.client.settings.poolSize": ConfigValue{
		1000,
		"number simultaneous active connections connections in a pool",
		1000,
		true, // immutable
	},
	"queryport.client.settings.poolOverflow": ConfigValue{
		30,
		"maximum number of connections in a pool",
		30,
		true, // immutable
	},
	"queryport.client.connPoolTimeout": ConfigValue{
		1000,
		"timeout, in milliseconds, is timeout for retrieving a connection " +
			"from the pool",
		1000,
		true, // immutable
	},
	"queryport.client.connPoolAvailWaitTimeout": ConfigValue{
		1,
		"timeout, in milliseconds, to wait for an existing connection " +
			"from the pool before considering the creation of a new one",
		1,
		true, // immutable
	},
	"queryport.client.retryScanPort": ConfigValue{
		2,
		"number of times to retry when scanport is not detectable",
		2,
		true, // immutable
	},
	"queryport.client.retryIntervalScanport": ConfigValue{
		10,
		"wait, in milliseconds, before re-trying for a scanport",
		10,
		true, // immutable
	},
	"queryport.client.servicesNotifierRetryTm": ConfigValue{
		1000,
		"wait, in milliseconds, before restarting the ServicesNotifier",
		1000,
		true, // immutable
	},
	// projector's adminport client, can be used by indexer.
	"indexer.projectorclient.retryInterval": ConfigValue{
		16,
		"retryInterval, in milliseconds when connection refused by server",
		16,
		true, // immutable
	},
	"indexer.projectorclient.maxRetries": ConfigValue{
		5,
		"maximum number of times to retry",
		5,
		true, // immutable
	},
	"indexer.projectorclient.exponentialBackoff": ConfigValue{
		2,
		"multiplying factor on retryInterval for every attempt with server",
		2,
		true, // immutable
	},
	"indexer.projectorclient.urlPrefix": ConfigValue{
		"/adminport/",
		"url prefix (script-path) for adminport used by projector",
		"/adminport/",
		true, // immutable
	},
	"indexer.adminPort": ConfigValue{
		"9100",
		"port for index ddl and status operations",
		"9100",
		true, // immutable
	},
	"indexer.scanPort": ConfigValue{
		"9101",
		"port for index scan operations",
		"9101",
		true, // immutable
	},
	"indexer.httpPort": ConfigValue{
		"9102",
		"port for external stats amd settings",
		"9102",
		true, // immutable
	},
	"indexer.streamInitPort": ConfigValue{
		"9103",
		"port for inital build stream",
		"9103",
		true, // immutable
	},
	"indexer.streamCatchupPort": ConfigValue{
		"9104",
		"port for catchup stream",
		"9104",
		true, // immutable
	},
	"indexer.streamMaintPort": ConfigValue{
		"9105",
		"port for maintenance stream",
		"9105",
		true, // immutable
	},
	"indexer.clusterAddr": ConfigValue{
		"127.0.0.1:8091",
		"Local cluster manager address",
		"127.0.0.1:8091",
		true, // immutable
	},
	"indexer.numVbuckets": ConfigValue{
		1024,
		"Number of vbuckets",
		1024,
		true, // immutable
	},
	"indexer.enableManager": ConfigValue{
		false,
		"Enable index manager",
		false,
		true, // immutable
	},
	"indexer.storage_dir": ConfigValue{
		"./",
		"Index file storage directory",
		"./",
		true, // immutable
	},
	"indexer.numSliceWriters": ConfigValue{
		1,
		"Number of Writer Threads for a Slice",
		1,
		true, // immutable
	},

	"indexer.sync_period": ConfigValue{
		uint64(2000),
		"Stream message sync interval in millis",
		uint64(2000),
		true, // immutable
	},

	"indexer.stats_cache_timeout": ConfigValue{
		uint64(5000),
		"Stats cache ttl in millis",
		uint64(5000),
		true, // immutable
	},

	// Indexer dynamic settings
	"indexer.settings.compaction.check_period": ConfigValue{
		1200000,
		"Compaction poll interval in seconds",
		1200000,
		false, // mutable
	},
	"indexer.settings.compaction.interval": ConfigValue{
		"00:00,00:00",
		"Compaction allowed interval",
		"00:00,00:00",
		false, // mutable
	},
	"indexer.settings.compaction.min_frag": ConfigValue{
		30,
		"Compaction fragmentation threshold percentage",
		30,
		false, // mutable
	},
	"indexer.settings.compaction.min_size": ConfigValue{
		uint64(1024 * 1024),
		"Compaction min file size",
		uint64(1024 * 1024),
		false, // mutable
	},
	"indexer.settings.persisted_snapshot.interval": ConfigValue{
		uint64(20000),
		"Persisted snapshotting interval in milliseconds",
		uint64(20000),
		false, // mutable
	},
	"indexer.settings.inmemory_snapshot.interval": ConfigValue{
		uint64(200),
		"InMemory snapshotting interval in milliseconds",
		uint64(200),
		false, // mutable
	},
	"indexer.settings.recovery.max_rollbacks": ConfigValue{
		5,
		"Maximum number of committed rollback points",
		5,
		false, // mutable
	},
	"indexer.settings.memory_quota": ConfigValue{
		uint64(256 * 1024 * 1024),
		"Maximum memory used by the indexer buffercache",
		uint64(256 * 1024 * 1024),
		false, // mutable
	},
	"indexer.settings.max_cpu_percent": ConfigValue{
		400,
		"Maximum percent of CPU that indexer can use. " +
			"EG, 200% in 4-core (400%) machine would set indexer to " +
			"use 2 cores",
		400,
		false, // mutable
	},
	"indexer.settings.log_level": ConfigValue{
		"debug",
		"Indexer logging level",
		"debug",
		false, // mutable
	},
	"indexer.settings.scan_timeout": ConfigValue{
		120000,
		"timeout, in milliseconds, timeout for index scan processing",
		120000,
		true, // immutable
	},

	"indexer.settings.send_buffer_size": ConfigValue{
		1024,
		"Buffer size for batching rows during scan result streaming",
		1024,
		true, // immutable
	},

	"indexer.settings.cpuProfFname": ConfigValue{
		"",
		"filename to dump cpu-profile for indexer.",
		"",
		false, // mutable
	},
	"indexer.settings.cpuProfile": ConfigValue{
		false,
		"boolean indicate whether to start or stop indexer cpu profiling.",
		false,
		false, // mutable
	},
	"indexer.settings.memProfFname": ConfigValue{
		"",
		"filename to dump mem-profile for indexer.",
		"",
		false, // mutable
	},
	"indexer.settings.memProfile": ConfigValue{
		false,
		"boolean to take current mem profile from indexer.",
		false,
		false, // mutable
	},

	"indexer.settings.maxVbQueueLength": ConfigValue{
		uint64(10000),
		"Maximum Length of Mutation Queue Per Vbucket. This " +
			"allocation is done per bucket.",
		uint64(10000),
		false, // mutable
	},

	"indexer.settings.largeSnapshotThreshold": ConfigValue{
		uint64(200),
		"Threshold For Considering a DCP Snapshot as Large. " +
			"Must be less than maxVbQueueLength.",
		uint64(200),
		false, // mutable
	},

	"indexer.settings.sliceBufSize": ConfigValue{
		uint64(50000),
		"Buffer for each slice to queue mutations before flush " +
			"to storage.",
		uint64(50000),
		false, // mutable
	},
	"indexer.settings.bufferPoolBlockSize": ConfigValue{
		16 * 1024,
		"Size of memory block in memory pool",
		16 * 1024,
		false,
	},
	"projector.settings.log_level": ConfigValue{
		"debug",
		"Projector logging level",
		"debug",
		false, // mutable
	},
}

// NewConfig from another
// Config object or from map[string]interface{} object
// or from []byte slice, a byte-slice of JSON string.
func NewConfig(data interface{}) (Config, error) {
	config := make(Config)
	err := config.Update(data)
	return config, err
}

// Update config object with data, can be a Config, map[string]interface{},
// []byte.
func (config Config) Update(data interface{}) error {
	fmsg := "CONF[] skipping setting key %q value '%v': %v"
	switch v := data.(type) {
	case Config: // Clone
		for key, value := range v {
			config.Set(key, value)
		}

	case []byte: // parse JSON
		m := make(map[string]interface{})
		if err := json.Unmarshal(v, &m); err != nil {
			return err
		}
		config.Update(m)

	case map[string]interface{}: // transform
		for key, value := range v {
			if cv, ok := SystemConfig[key]; ok { // valid config.
				if _, ok := config[key]; !ok {
					config[key] = cv // copy by value
				}
				if err := config.SetValue(key, value); err != nil {
					logging.Warnf(fmsg, key, value, err)
				}

			} else {
				logging.Errorf("invalid config param %q", key)
			}
		}

	default:
		return nil
	}
	return nil
}

// Clone a new config object.
func (config Config) Clone() Config {
	clone := make(Config)
	for key, value := range config {
		clone[key] = value
	}
	return clone
}

// Override will clone `config` object and update parameters with
// values from `others` instance. Will skip immutable fields.
func (config Config) Override(others ...Config) Config {
	newconfig := config.Clone()
	for _, other := range others {
		for key, cv := range other {
			if newconfig[key].Immutable { // skip immutables.
				continue
			}
			ocv, ok := newconfig[key]
			if !ok {
				ocv = cv
			} else {
				ocv.Value = cv.Value
			}
			config[key] = ocv
		}
	}
	return config
}

// OverrideForce will clone `config` object and update parameters with
// values from `others` instance. Will force override immutable fields
// as well.
func (config Config) OverrideForce(others ...Config) Config {
	newconfig := config.Clone()
	for _, other := range others {
		for key, cv := range other {
			ocv, ok := newconfig[key]
			if !ok {
				ocv = cv
			} else {
				ocv.Value = cv.Value
			}
			config[key] = ocv
		}
	}
	return config
}

// LogConfig will check wether a configuration parameter is
// mutable and log that information.
func (config Config) LogConfig(prefix string) {
	for key, cv := range config {
		if cv.Immutable {
			fmsg := "%v immutable settings %v cannot be update to `%v`\n"
			logging.Warnf(fmsg, prefix, key, cv.Value)
		} else {
			fmsg := "%v settings %v will updated to `%v`\n"
			logging.Infof(fmsg, prefix, key, cv.Value)
		}
	}
}

// SectionConfig will create a new config object with parameters
// starting with `prefix`. If `trim` is true, then config
// parameter will be trimmed with the prefix string.
func (config Config) SectionConfig(prefix string, trim bool) Config {
	section := make(Config)
	for key, value := range config {
		if strings.HasPrefix(key, prefix) {
			if trim {
				section[strings.TrimPrefix(key, prefix)] = value
			} else {
				section[key] = value
			}
		}
	}
	return section
}

func (config Config) FilterConfig(subs string) Config {
	newConfig := make(Config)
	for key, value := range config {
		if strings.Contains(key, subs) {
			newConfig[key] = value
		}
	}
	return newConfig
}

// Set ConfigValue for parameter. Mutates the config object.
func (config Config) Set(key string, cv ConfigValue) Config {
	config[key] = cv
	return config
}

// SetValue config parameter with value. Mutates the config object.
func (config Config) SetValue(key string, value interface{}) error {
	cv, ok := config[key]
	if !ok {
		return errors.New("invalid config parameter")
	}

	defType := reflect.TypeOf(cv.DefaultVal)
	valType := reflect.TypeOf(value)

	if valType.ConvertibleTo(defType) {
		v := reflect.ValueOf(value)
		v = reflect.Indirect(v)
		value = v.Convert(defType).Interface()
		valType = defType
	}

	if defType != reflect.TypeOf(value) {
		return fmt.Errorf("%v: Value type mismatch, %v != %v (%v)",
			key, valType, defType, value)
	}
	cv.Value = value
	config[key] = cv

	return nil
}

// Json will marshal config into JSON string.
func (config Config) Json() []byte {
	kvs := make(map[string]interface{})
	for key, value := range config {
		kvs[key] = value.Value
	}

	bytes, _ := json.Marshal(kvs)
	return bytes
}

// Int assumes config value is an integer and returns the same.
func (cv ConfigValue) Int() int {
	if val, ok := cv.Value.(int); ok {
		return val
	} else if val, ok := cv.Value.(float64); ok {
		return int(val)
	}
	panic(fmt.Errorf("not support Int() on %v", cv))
}

// Uint64 assumes config value is 64-bit integer and returns the same.
func (cv ConfigValue) Uint64() uint64 {
	return cv.Value.(uint64)
}

// String assumes config value is a string and returns the same.
func (cv ConfigValue) String() string {
	return cv.Value.(string)
}

// Strings assumes config value is comma separated string items.
func (cv ConfigValue) Strings() []string {
	ss := make([]string, 0)
	for _, s := range strings.Split(cv.Value.(string), ",") {
		s = strings.Trim(s, " \t\r\n")
		if len(s) > 0 {
			ss = append(ss, s)
		}
	}
	return ss
}

// Bool assumes config value is a Bool and returns the same.
func (cv ConfigValue) Bool() bool {
	return cv.Value.(bool)
}
